python/parquet_tools/common/schema.py

```diff
@@ -0,0 +1,700 @@
+            raise ParmError(6, 'Schema file', schfile, 'not found')
+
+        # Open the schfile and get all the records
+        #
+        try:
+            with open(schfile) as sch_file:
+                recs = [line.rstrip('\n') for line in sch_file]
+        except OSError as ex:
+            raise FatalError(ex, 'Unable to open schema file', schfile)
+
+        # Run through each record grabbing the first two tokens (the second one
+        # may be missing).
+        #
+        for line in recs:
+            toks = line.split()
+            if len(toks) > 0:
+                Schema._add_col(toks[0])
+                if not do_auto and len(toks) > 1:
+                    nullok = Schema._can_be_null(toks[2:])
+                    if not nullok:
+                        Schema._colNotNull.append(toks[0])
+                    Schema._get_type(toks[0], toks[1], nullok, colnulls)
+                elif toks[0]:
+                    Schema._colSpecs.append([toks[0], 'auto', 'auto'])
+
+        # Record original number of colums
+        #
+        Schema.colOrigs = len(Schema.colNames)
+        Schema._cNameMaw = Schema._cNameMax
+
+        # Check if we should augment the schema for integers that may be null
+        #
+        if colnulls:
+            for x in colnulls:
+                Schema.colNVChk.append(x[0])
+                Schema._add_col(x[1])
+
+    # *************************************************************************
+    # *                        _ r e v e r t 2 n u l l                        *
+    # *************************************************************************
+    @classmethod
+    def _revert2null(cls, df, nilval):
+        """Revert integer fields that originally cotained null to be so.
+
+        Parameters
+        ----------
+        df : 'PANDAS dataframe'
+            This is the dataframe which is to be manipulated into submission.
+        nilval : 'string'
+            This is the sequence that defines a "null" value.
+
+        Raises
+        ------
+        FatalError
+        """
+
+        # If there are no null values at all, simply return.
+        #
+        if not Schema.colNVChk:
+            return None
+
+        # For each column that may have a null value, convert it to a string.
+        # Then replace it's value with a null sequence if it should contain it.
+        # Note that we need to get around flake8 as pandas only allows
+        # comparisons to values and we need to base execution on a boolean.
+        #
+        trueval = True
+        for cnum in Schema.colNVChk:
+            cname = Schema.colNames[cnum]
+            try:
+                df[cname] = df[cname].apply(str)
+            except Exception as ex:
+                raise FatalError(ex, 'Unable to convert integer column',
+                                 cname, 'to string')
+            cisnil = cname + Schema._nilSuffx
+            df.loc[df[cisnil] == trueval, cname] = nilval
```

Comment:
this line is unreadable. please change it, and use named variables or something.

Reply:
Perhaps because you are not familiar with dataframe vector programming. This is how it's done. Essentially, we are doing a column value replacement on all rows that satisfy a particular condition. Perhaps you mean that I don't use sentences for the variable names. Given that cisnil read as "c is nil" is immediately defined above and not used elsewhere should be clear. "cname" (used consistently) is short hand for columnname which I don't find particularly readable . I suppose "colname" may be a good compromise.